### PR TITLE
2.3 - Testing - Andrew T. Poe

### PR DESCRIPTION
**Pull request by andrewtpoe**
All functions do not have tests written for them, but this commit currently demonstrates several techniques of testing.

### CODE REVIEW

File: App.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+# OK, lets initialize this ATM with a $10,000 balance
+atm = Machine.new
+atm.balance = 10000
+
+
+# # This line creates a new instance of the Program.
+program = Program.new(atm)
+
+# This executes the program until the user has decided to exit.
+while program.exit_prog == false
```
Comment (rpearce) [inline App.rb:13]: would `until` make more sense here?

File: classes/machine.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# The machine class provides the information on the actual ATM
+class Machine
+  attr_accessor :balance
+
+  # Verify that the ATM contains sufficient money for the withdrawal
+  def can_withdraw?(amount)
+    amount.class == Fixnum && @balance >= amount && amount > 0
```
Comment (rpearce) [inline classes/machine.rb:7]: I would put `amount > 0` before checking the balance (which might be a costly procedure). More in depth answer here:
https://github.com/easmith2/2.3_testing/pull/1#discussion_r33641217

File: classes/program.rb [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        data = [user.name, user.pin, user.balance]
+        csv << data
+      end
+    end
+  end
+
+  # Extract gets.chomp for easier testing purposes.
+  def chomp
+    var = gets.chomp
+    return var
```
Comment (rpearce) [inline classes/program.rb:245]: `return` is not needed here :)

File: classes/program.rb [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+      @account_holders.each do |user|
+        data = [user.name, user.pin, user.balance]
+        csv << data
+      end
+    end
+  end
+
+  # Extract gets.chomp for easier testing purposes.
+  def chomp
+    var = gets.chomp
```
Comment (rpearce) [inline classes/program.rb:244]: what happens if they put in empty spaces? Perhaps a `.strip` call would be useful here

File: classes/user.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  def initialize
+    @name = ""
+    @pin = ""
+    @balance = 0
+  end
+
+  # Verify that the user can withdraw the requested amount
+  def can_withdraw?(amount)
+    amount.class == Fixnum && @balance >= amount && amount > 0
+  end
```
Comment (rpearce) [inline classes/user.rb:14]: given this is in 2 places, I would consider using a `module`/mixin to share code

File: tests.rb [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    menu_sel = nil
+    @p.stub(:get_input, 1) do
+      menu_sel = @p.get_menu_selection(menu_arr)
+    end
+    assert_equal 1, menu_sel
+  end
+
+  # Test get_user(login)
+
+  # Test show_acc_menu
```
Comment (rpearce) [inline tests.rb:120]: I would make these methods with a `skip "this is why..."` method call within them so they show up as skipped tests

**Comment by rpearce** [comment]
Groovy. Nice tests

File: classes/program.rb (hunk 4 above)
Comment (andrewtpoe) [inline classes/program.rb:244]: for something as secure as an ATM should be, I felt it was better to have an exact match based on user input. For general form fields, a .strip call would likely be in order.

**Pull request merged**